Label containers with stack name when deploying stack/bundle

yongtang opened the pull request:
**\- What I did**

This fix tries to address the issue raised in #24881 where the stack name is not visable when inspecting containers deployed through stacks.

**\- How I did it**

This fix adds `labelNamespace` (`com.docker.stack.namespace`) to containers at the client side so that the stack name label will show up after containers are deployed.

**\- How to verify it**

This fix is tested manually:
- Build the binary with `make DOCKER_EXPERIMENTAL=1`
- Create a bundle file and deploy with the `docker stack deploy`
- Verify that `com.docker.stack.namespace` has been set properly by checking contaners with `docker inspect <ContainerID>`

**\- Description for the changelog**

**\- A picture of a cute animal (not mandatory but encouraged)**

This fix fixes #24881.

Signed-off-by: Yong Tang yong.tang.github@outlook.com

thaJeztah commented:
/cc @dnephin

dnephin commented on api/client/stack/common.go:
```diff
@@ -12,7 +12,9 @@ import (
-	labelNamespace = "com.docker.stack.namespace"
+	// TODO: The `experimental` should be removed once the
+	// stack feature is out of the experimental status
+	labelNamespace = "com.docker.experimental.stack.namespace"
```
If we make this change it's going to be harder for users to transition off of experimental. I think we should keep the old label.

yongtang replied:
Thanks @dnephin the pull request has been updated.

dnephin commented on daemon/cluster/executor/container/container.go:
```diff
@@ -139,6 +139,11 @@ func (c *containerConfig) labels() map[string]string {
+	// we apply service level labels, which could be obtained from ServiceAnnotations.Labels
+	for k, v := range c.task.ServiceAnnotations.Labels {
+		labels[k] = v
+	}
```
There has been a lot of discussion about  this, and so far the agreement is that we don't want to copy the labels... yet. It needs more discussion apparently.

So instead of copying the label, we can apply a container label from the client , in the same place we add the service label.

yongtang replied:
Thanks @dnephin the pull request has been updated. Now the labels are added at the client side to the containers.

dnephin commented on api/client/stack/deploy.go:
```diff
@@ -168,6 +168,10 @@ func deployServices(
+					// Service Labels will not be copied to Containers
+					// automatically during the deployment so we apply
+					// it here.
+					Labels: getStackLabels(namespace, service.Labels),
```
This should pass `nil` as the second arg, otherwise this copies all labels.

yongtang replied:
Thanks @dnephin for pointing that out. The pull request has been updated. Let me know if there are any issues.

yongtang committed:
Label containers with stack name when deploying stack/bundle

This fix tries to address the issue raised in 24881 where
the stack name is not visable when inspecting containers
deployed through stacks.

This fix adds `labelNamespace` (`com.docker.stack.namespace`)
to containers at the client side so that the stack name label
will show up after containers are deployed.

This fix is tested manually:
 - Build the binary with `make DOCKER_EXPERIMENTAL=1`
 - Create a bundle file and deploy with the `docker stack deploy`
 - Verify that `com.docker.stack.namespace` has been
set properly by checking contaners with `docker inspect <ContainerID>`

This fix fixes 24881.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

dnephin commented:
LGTM

crosbymichael commented:
LGTM
